generators/E906LegacyVtxGen/SQPrimaryVertexGen.C

```diff
@@ -0,0 +1,270 @@
+
+ 
+  beamProfile = new TF2("beamProfile", &SQPrimaryVertexGen::funcBeamProfile, -6.,6.,-6.,6.,4,2);
+  beamProfile->SetNpx(200);//SetNpx and Npy for smooth distribution (default is 30)
+  beamProfile->SetNpy(200); 
+ 
+  beamProfile->SetParameter(0, rc->get_DoubleFlag("X_BEAM"));
+  beamProfile->SetParameter(1, rc->get_DoubleFlag("SIGX_BEAM"));
+  beamProfile->SetParameter(2, rc->get_DoubleFlag("Y_BEAM"));
+  beamProfile->SetParameter(3, rc->get_DoubleFlag("SIGY_BEAM"));
+
+  rndm.SetSeed(PHRandomSeed());
+
+  //If inited from a root file, then skip this step, so that it can be used independent of the Fun4All
+  if(geoManager == nullptr) geoManager = PHGeomUtility::GetTGeoManager(topNode);
+
+  //Read the default 
+  defaultMatProf = new MaterialProfile;
+  fillMaterialProfile(defaultMatProf, 0., 0.);
+
+  inited = true;
+}
+
+void SQPrimaryVertexGen::fillMaterialProfile(MaterialProfile* prof, double xvtx, double yvtx)
+{
+  //std::cout << xvtx << "  " << yvtx << std::endl;
+  //retrieve the z position of all geometry boundaries
+  recoConsts* rc = recoConsts::instance();
+
+  double z[100];
+  int nBoundaries = 0;
+
+  double eps = 1.E-6;
+  double z_curr = -800.;//not beyond collimator(z= -602.36 cm,lenth=121.92 cm) @Abi //-500.;
+  z[0] = z_curr;
+  while(z_curr < 503.)
+    {
+      geoManager->IsSameLocation(xvtx, yvtx, z_curr, true);
+      geoManager->SetCurrentDirection(0., 0., 1.);
+
+      TGeoNode* node = geoManager->FindNextBoundary(600.);
+      double z_step  = geoManager->GetStep();
+
+      if(z_step > eps)
+	{
+	  z_curr += z_step;
+	  z[++nBoundaries] = z_curr; 
+	}
+      else
+	{
+	  z_curr += eps;
+	}
+    }
+
+  //use the z boundaries to create interactibles
+  for(int i = 0; i < nBoundaries; ++i)
+    {
+      double z_center = 0.5*(z[i] + z[i+1]);
+      if(z_center > 503.) continue;
+ 
+      SQBeamlineObject obj(geoManager->FindNode(xvtx, yvtx, z_center)->GetVolume()->GetMaterial());
+      TString name = geoManager->FindNode(xvtx, yvtx, z_center)->GetVolume()->GetName();
+      obj.z_up = z[i];
+      obj.z_down = z[i+1];
+      obj.z0 = z_center;
+      obj.length = z[i+1] - z[i];
+    
+      //TString name = obj.name;    
+      //std::cout<<" Name of beam volume "<<name<<std::endl;  
+
+      if (_targetOnlyMode){
+	if(! name.Contains("Target")) continue;
+      }
+
+      if (_dumpOnlyMode){
+	if(! name.Contains("fmag")) continue;
+      }
+
+      prof->interactables.push_back(obj);
+    }
+
+  //calculate all related properties
+  prof->calcProb();
+
+  // for(int i = 0; i < prof->nPieces; ++i)
+  // {
+  //   std::cout << prof->interactables[i] << std::endl;
+  // }
+  // return 0;
+}
+
+
+TVector3 SQPrimaryVertexGen::generateVertex()
+{
+  if(!inited) init();
+
+  double xvtx, yvtx, zvtx;
+  generateVtxPerp(xvtx, yvtx);
+
+  bool inTarget = ((xvtx-targetX0)*(xvtx-targetX0)/targetSX/targetSX + (yvtx-targetY0)*(yvtx-targetX0)/targetSY/targetSY) < 1.;
+  MaterialProfile* activeProf;
+  if(!inTarget)
```

Comment:
What is the purpose of this "if" condition?  I understand that the material profile at x = y = 0 (i.e. `defaultMatProf`) is used when the beam position generated is inside the target area.  But I wonder why such a special handling is necessary.

Reply:
I assume @liuk implemented this condition to save time to fill the material profile inside the target area as we don't expect to change any geometry in that region. In principle, it should work even if each time we fill the material profile corresponding to (x, y) values. But @liuk  might can give the exact answer for this.

Reply:
If we keep this, could you add a comment to the code that explains the purpose and the assumption (the material profile inside the target area must be uniform)?